### PR TITLE
Add ast-grep as an engine 

### PR DESCRIPTION
**Pull request by Harsh24893**
# This Diff 
**Summary**
- In commit 1, adding no-op ast-grep engine to codemod cli 
- In commit 2, added implementation of ast-grep engine to cli 

**Testing**

I put the following code in a file called abc.js in folder `/Users/harshgupta/Desktop/src/codemod/apps/cli/ast/abc.js`
```
import { A, B, C } from './barrel';

function Component() {
    const [name, setName] = useState < string > ('React')
}
``` 

In directory `/Users/harshgupta/.codemod`, chose a random folder `XG6X5ceI2dzkLW7CoBI0aD3W_7Q`, and updated it's config.json to have engine `ast-grep`. 

config.json after update 
```
"schemaVersion":"1.0.0","engine":"ast-grep","arguments":[],"name":"cal.com/app-directory-boilerplate-calcom"}
```

Also added a file called rule.yaml in `XG6X5ceI2dzkLW7CoBI0aD3W_7Q` and content of yaml file was 
```
id: change_js
language: javascript
rule:
  pattern: import {$$$IDENTS} from './barrel'
rewriters:
- id: rewrite-identifer
  rule:
    pattern: $IDENT
    kind: identifier
  fix: import $IDENT from './barrel/$IDENT'
transform:
  IMPORTS:
    rewrite:
      rewriters: [rewrite-identifer]
      source: $$$IDENTS
      joinBy: "\n"
fix: $IMPORTS
```

Built the codemod using 
```npm run build && npm link```

Executed 
```codemod cal.com/app-directory-boilerplate-calcom``` with useCache true and --target=/Users/harshgupta/Desktop/src/codemod/apps/cli/ast 

**useCache set to true so that it doesn't redownload the codemod from S3 which would update config.json and overwrite the engine.

**Resources**
https://ast-grep.github.io/guide/introduction.html 

# Follow up 
- Come up with some codemod for ast-grep and publish to registry 
- Add a script to execute ast-grep codemods in a batch manner and create PRs from it

### CODE REVIEW

**Comment by Harsh24893** [comment]
I am seeing some build failures as it is not able to figure out js-yaml files, but I don't see failures while executing codemod.

File: apps/cli/src/downloadCodemod.ts [hunk 1]
```diff
@@ -111,6 +111,25 @@ export class CodemodDownloader implements CodemodDownloaderBlueprint {
...
+				yamlPath,
+			);
+
+			await writeFile(yamlPath, yamlData);
+
+			return {
+				source: 'registry',
+				name,
+				engine: config.engine,
+				yamlPath: yamlPath,
```
Comment (r4zendev) [inline apps/cli/src/downloadCodemod.ts:127]: use shorthand

File: apps/cli/src/runAstgrepCodemod.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+				// Recursively call the function for subdirectories
+				await iterateFiles(entryPath);
+			} else if (entry.isFile()) {
+				// If the entry is a file, log its path
+				const fileExtension = path.extname(entryPath).slice(1);
+				if (fileExtension !== language) {
+					continue;
+				}
+				const astCommand = `sg scan -r ${rulePath} ${entryPath} -U`;
+				execSync(astCommand);
```
Comment (r4zendev) [inline apps/cli/src/runAstgrepCodemod.ts:41]: i'm not sure this is going to work unless user has installed ast-grep on his machine

File: apps/cli/src/downloadCodemod.ts [hunk 3]
```diff
@@ -111,6 +111,25 @@ export class CodemodDownloader implements CodemodDownloaderBlueprint {
 			}
 		}
 
+		if (config.engine === 'ast-grep') {
+			const yamlPath = join(directoryPath, 'rule.yaml');
+			const yamlData = await this._fileDownloadService.download(
```
Comment (DmytroHryshyn) [inline apps/cli/src/downloadCodemod.ts:116]: can it throw? Should we wrap it in try/catch?

File: apps/cli/src/downloadCodemod.ts [hunk 4]
```diff
@@ -111,6 +111,25 @@ export class CodemodDownloader implements CodemodDownloaderBlueprint {
...
+				`${CODEMOD_REGISTRY_URL}/${hashDigest}/rule.yaml`,
+				yamlPath,
+			);
+
+			await writeFile(yamlPath, yamlData);
```
Comment (DmytroHryshyn) [inline apps/cli/src/downloadCodemod.ts:121]: I think `download` method already saves file to the filesystem

File: apps/cli/src/runAstgrepCodemod.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+	const yamlObject = yaml.load(yamlString);
+	const language = languageToExtension(yamlObject.language);
+	printer.printConsoleMessage(
+		'info',
+		`Executing ast-grep for language : ${language}`,
+	);
+
+	// Function to recursively iterate over files in the directory
+	const iterateFiles = async (dirPath: string) => {
+		const entries = await fs.promises.readdir(dirPath, {
```
Comment (DmytroHryshyn) [inline apps/cli/src/runAstgrepCodemod.ts:23]: do we need to take into account `include` and `exclude` patterns here (passed as cli args)?

Comment (DmytroHryshyn) [reply apps/cli/src/runAstgrepCodemod.ts:23]: maybe we can use `glob` lib for this use case?

File: apps/cli/src/runAstgrepCodemod.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+			if (entry.isDirectory()) {
+				// Recursively call the function for subdirectories
+				await iterateFiles(entryPath);
+			} else if (entry.isFile()) {
+				// If the entry is a file, log its path
+				const fileExtension = path.extname(entryPath).slice(1);
+				if (fileExtension !== language) {
+					continue;
+				}
+				const astCommand = `sg scan -r ${rulePath} ${entryPath} -U`;
```
Comment (DmytroHryshyn) [inline apps/cli/src/runAstgrepCodemod.ts:40]: will this work on `Windows`? i think we need to add quotes to the args to make it work in Windows cmd. I will check later.

File: apps/cli/src/runAstgrepCodemod.ts (hunk 2 above)
Comment (r4zendev) [reply apps/cli/src/runAstgrepCodemod.ts:41]: also we don't need to execute changes here, we need to build new files, then execute buildFileCommands() and pass it up. similar how it's done for other engines. not sure if you're doing the same here.

File: apps/cli/src/runAstgrepCodemod.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+import { PrinterBlueprint } from './printer.js';
+
+export const runAstgrep = async (
+	printer: PrinterBlueprint,
+	rulePath: string,
+	targetDirectory: string,
+): Promise<void> => {
+	const yamlString = await readFile(rulePath, { encoding: 'utf8' });
+	const yamlObject = yaml.load(yamlString);
+	const language = languageToExtension(yamlObject.language);
```
Comment (DmytroHryshyn) [inline apps/cli/src/runAstgrepCodemod.ts:15]: lets call this variable `extension`

**Comment by mohab-sameh** [comment]
@all-contributors add @Harsh24893 for code

File: apps/cli/src/downloadCodemod.ts (hunk 3 above)
Comment (Harsh24893) [reply apps/cli/src/downloadCodemod.ts:116]: It shouldn't throwing, but adding a try catch is always a good idea. Will do it.

File: apps/cli/src/runAstgrepCodemod.ts (hunk 5 above)
Comment (Harsh24893) [reply apps/cli/src/runAstgrepCodemod.ts:24]: I was thinking of shipping this and doing other changes once Sasha refactors the cli.

File: apps/cli/src/runAstgrepCodemod.ts (hunk 6 above)
Comment (Harsh24893) [reply apps/cli/src/runAstgrepCodemod.ts:41]: Updated the code

File: apps/cli/src/runAstgrepCodemod.ts (hunk 2 above)
Comment (Harsh24893) [reply apps/cli/src/runAstgrepCodemod.ts]: I think just installing it for the user is a good idea to get ahead. Also this ll be in beta mode, so we can work on better devX as we move ahead. 

What do you mean by build new files.

File: apps/cli/src/runAstgrepCodemod.ts [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+// Function to check if ast-grep is available or not
+const installSgCommandIfNotAvailable = (printer: PrinterBlueprint): void => {
+	try {
+		// Use `which` command to check if the command is available
+		execSync(`which sg`);
+	} catch (error) {
+		// If `which` command fails, the command is not available
+		printer.printConsoleMessage(
+			'info',
+			'ast-grep is not avialable, installing it globally',
```
Comment (r4zendev) [inline apps/cli/src/runAstgrepCodemod.ts:78]: available

**Review by r4zendev — DISMISSED** [review]

**Review by r4zendev — APPROVED** [review]

**Pull request merged**